### PR TITLE
[new release] lsp (3 packages) (1.20.0)

### PR DESCRIPTION
**Pull request by voodoos**
CHANGES:

## Features

- Add custom [`ocamllsp/typeSearch`](/ocaml-lsp-server/docs/ocamllsp/typeSearch-spec.md) request (ocaml/ocaml-lsp#1369)

- Make MerlinJump code action configurable (ocaml/ocaml-lsp#1376)

- Add custom [`ocamllsp/jump`](/ocaml-lsp-server/docs/ocamllsp/merlinJump-spec.md) request (ocaml/ocaml-lsp#1374)

## Fixes

- Fix fd leak in running external processes for preprocessing (ocaml/ocaml-lsp#1349)

- Fix prefix parsing for completion of object methods (ocaml/ocaml-lsp#1363, fixes ocaml/ocaml-lsp#1358)

- Remove some duplicates in the `selectionRange` answers (ocaml/ocaml-lsp#1368)

### CODE REVIEW

File: packages/ocaml-lsp-server/ocaml-lsp-server.1.20.0/opam [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  "lsp" {= version}
+  "jsonrpc" {= version}
+  "re" {>= "1.5.0"}
+  "ppx_yojson_conv_lib" {>= "v0.14"}
+  "dune-rpc" {>= "3.4.0"}
+  "chrome-trace" {>= "3.3.0"}
+  "dyn"
+  "stdune"
+  "fiber" {>= "3.1.1" & < "4.0.0"}
+  "ocaml" {>= "5.2.0"}
```
Comment (mseri) [inline packages/ocaml-lsp-server/ocaml-lsp-server.1.20.0/opam:33]:
```suggestion
  "ocaml" {>= "5.2.0" & < "5.3~"}
```
due to
```

#=== ERROR while compiling ocaml-lsp-server.1.20.0 ============================#
# context              2.4.0~alpha1~dev | linux/x86_64 | ocaml-base-compiler.5.3.0~beta1 | pinned(https://github.com/ocaml/ocaml-lsp/releases/download/1.20.0/lsp-1.20.0.tbz)
# path                 ~/.opam/5.3~beta1/.opam-switch/build/ocaml-lsp-server.1.20.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ocaml-lsp-server -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/ocaml-lsp-server-7-ead90a.env
# output-file          ~/.opam/log/ocaml-lsp-server-7-ead90a.out
### output ###
# (cd _build/default/ocaml-lsp-server/vendor/lev && /usr/bin/gcc -O2 -fno-strict-aliasing -fwrapv -fPIC -pthread -D_FILE_OFFSET_BITS=64 -g -I /home/opam/.opam/5.3~beta1/lib/ocaml -I /home/opam/.opam/5.3~beta1/lib/ocaml/unix -o ev.o -c ev.c)
# submodules/lev/lev/vendor/ev.c:2143:31: warning: 'ev_default_loop_ptr' initialized and declared 'extern'
# (cd _build/default && /home/opam/.opam/5.3~beta1/bin/ocamlopt.opt -w -40 -alert -unstable -w -58 -g -I ocaml-lsp-server/src/.ocaml_lsp_server.objs/byte -I ocaml-lsp-server/src/.ocaml_lsp_server.objs/native -I /home/opam/.opam/5.3~beta1/lib/base -I /home/opam/.opam/5.3~beta1/lib/chrome-trace -I /home/opam/.opam/5.3~beta1/lib/csexp -I /home/opam/.opam/5.3~beta1/lib/dune-build-info -I /home/opam/.opam/5.3~beta1/lib/dune-rpc -I /home/opam/.opam/5.3~beta1/lib/dune-rpc/private -I /home/opam/.opam/5.3~beta1/lib/dyn -I /home/opam/.opam/5.3~beta1/lib/fiber -I /home/opam/.opam/5.3~beta1/lib/jsonrpc -I /home/opam/.opam/5.3~beta1/lib/lsp -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/analysis -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/commands -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/dot_protocol -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/extend -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/kernel -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_merlin_specific -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_parsing -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_typing -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_utils -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/query_commands -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/query_protocol -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/utils -I /home/opam/.opam/5.3~beta1/lib/ocaml/unix -I /home/opam/.opam/5.3~beta1/lib/ocamlc-loc -I /home/opam/.opam/5.3~beta1/lib/ocamlformat-rpc-lib -I /home/opam/.opam/5.3~beta1/lib/ordering -I /home/opam/.opam/5.3~beta1/lib/pp -I /home/opam/.opam/5.3~beta1/lib/ppx_yojson_conv_lib -I /home/opam/.opam/5.3~beta1/lib/re -I /home/opam/.opam/5.3~beta1/lib/spawn -I /home/opam/.opam/5.3~beta1/lib/stdune -I /home/opam/.opam/5.3~beta1/lib/stdune/filesystem_stubs -I /home/opam/.opam/5.3~beta1/lib/xdg -I /home/opam/.opam/5.3~beta1/lib/yojson -I lsp-fiber/src/.lsp_fiber.objs/byte -I lsp-fiber/src/.lsp_fiber.objs/native -I ocaml-lsp-server/vendor/cmarkit/.cmarkit.objs/byte -I ocaml-lsp-server/vendor/cmarkit/.cmarkit.objs/native -I ocaml-lsp-server/vendor/lev-fiber-csexp/.lev_fiber_csexp.objs/byte -I ocaml-lsp-server/vendor/lev-fiber-csexp/.lev_fiber_csexp.objs/native -I ocaml-lsp-server/vendor/lev-fiber/.lev_fiber.objs/byte -I ocaml-lsp-server/vendor/lev-fiber/.lev_fiber.objs/native -I ocaml-lsp-server/vendor/lev/.lev.objs/byte -I ocaml-lsp-server/vendor/lev/.lev.objs/native -I vendor/odoc-parser/src/.odoc_parser.objs/byte -I vendor/odoc-parser/src/.odoc_parser.objs/native -intf-suffix .ml -no-alias-deps -open Ocaml_lsp_server__ -o ocaml-lsp-server/src/.ocaml_lsp_server.objs/native/ocaml_lsp_server__Action_extract.cmx -c -impl ocaml-lsp-server/src/code_actions/action_extract.ml)
# File "ocaml-lsp-server/src/code_actions/action_extract.ml", line 78, characters 8-32:
# 78 |       | Texp_match (_, cases, _) -> found_if_case_contains cases
#              ^^^^^^^^^^^^^^^^^^^^^^^^
# Error: The constructor "Texp_match" expects 4 argument(s),
#        but is applied here to 3 argument(s)
# (cd _build/default && /home/opam/.opam/5.3~beta1/bin/ocamlopt.opt -w -40 -alert -unstable -w -58 -g -I ocaml-lsp-server/src/.ocaml_lsp_server.objs/byte -I ocaml-lsp-server/src/.ocaml_lsp_server.objs/native -I /home/opam/.opam/5.3~beta1/lib/base -I /home/opam/.opam/5.3~beta1/lib/chrome-trace -I /home/opam/.opam/5.3~beta1/lib/csexp -I /home/opam/.opam/5.3~beta1/lib/dune-build-info -I /home/opam/.opam/5.3~beta1/lib/dune-rpc -I /home/opam/.opam/5.3~beta1/lib/dune-rpc/private -I /home/opam/.opam/5.3~beta1/lib/dyn -I /home/opam/.opam/5.3~beta1/lib/fiber -I /home/opam/.opam/5.3~beta1/lib/jsonrpc -I /home/opam/.opam/5.3~beta1/lib/lsp -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/analysis -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/commands -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/dot_protocol -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/extend -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/kernel -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_merlin_specific -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_parsing -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_typing -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_utils -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/query_commands -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/query_protocol -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/utils -I /home/opam/.opam/5.3~beta1/lib/ocaml/unix -I /home/opam/.opam/5.3~beta1/lib/ocamlc-loc -I /home/opam/.opam/5.3~beta1/lib/ocamlformat-rpc-lib -I /home/opam/.opam/5.3~beta1/lib/ordering -I /home/opam/.opam/5.3~beta1/lib/pp -I /home/opam/.opam/5.3~beta1/lib/ppx_yojson_conv_lib -I /home/opam/.opam/5.3~beta1/lib/re -I /home/opam/.opam/5.3~beta1/lib/spawn -I /home/opam/.opam/5.3~beta1/lib/stdune -I /home/opam/.opam/5.3~beta1/lib/stdune/filesystem_stubs -I /home/opam/.opam/5.3~beta1/lib/xdg -I /home/opam/.opam/5.3~beta1/lib/yojson -I lsp-fiber/src/.lsp_fiber.objs/byte -I lsp-fiber/src/.lsp_fiber.objs/native -I ocaml-lsp-server/vendor/cmarkit/.cmarkit.objs/byte -I ocaml-lsp-server/vendor/cmarkit/.cmarkit.objs/native -I ocaml-lsp-server/vendor/lev-fiber-csexp/.lev_fiber_csexp.objs/byte -I ocaml-lsp-server/vendor/lev-fiber-csexp/.lev_fiber_csexp.objs/native -I ocaml-lsp-server/vendor/lev-fiber/.lev_fiber.objs/byte -I ocaml-lsp-server/vendor/lev-fiber/.lev_fiber.objs/native -I ocaml-lsp-server/vendor/lev/.lev.objs/byte -I ocaml-lsp-server/vendor/lev/.lev.objs/native -I vendor/odoc-parser/src/.odoc_parser.objs/byte -I vendor/odoc-parser/src/.odoc_parser.objs/native -intf-suffix .ml -no-alias-deps -open Ocaml_lsp_server__ -o ocaml-lsp-server/src/.ocaml_lsp_server.objs/native/ocaml_lsp_server__Folding_range.cmx -c -impl ocaml-lsp-server/src/folding_range.ml)
# File "ocaml-lsp-server/src/folding_range.ml", lines 139-162, characters 6-60:
# 139 | ......match p.ppat_desc with
# 140 |       | Ppat_record (bdgs, _) ->
# 141 |         Range.of_loc p.ppat_loc |> push;
# 142 |         List.iter bdgs ~f:(fun (lident, pat) ->
# 143 |           let lident_range = Range.of_loc lident.Asttypes.loc in
# ...
# 159 |       | Ppat_exception _
# 160 |       | Ppat_extension _
# 161 |       | Ppat_open _
# 162 |       | Ppat_any -> Ast_iterator.default_iterator.pat self p
# Warning 8 [partial-match]: this pattern-matching is not exhaustive.
# Here is an example of a case that is not matched:
# Ppat_effect (_, _)
# (cd _build/default && /home/opam/.opam/5.3~beta1/bin/ocamlopt.opt -w -40 -alert -unstable -w -58 -g -I ocaml-lsp-server/src/.ocaml_lsp_server.objs/byte -I ocaml-lsp-server/src/.ocaml_lsp_server.objs/native -I /home/opam/.opam/5.3~beta1/lib/base -I /home/opam/.opam/5.3~beta1/lib/chrome-trace -I /home/opam/.opam/5.3~beta1/lib/csexp -I /home/opam/.opam/5.3~beta1/lib/dune-build-info -I /home/opam/.opam/5.3~beta1/lib/dune-rpc -I /home/opam/.opam/5.3~beta1/lib/dune-rpc/private -I /home/opam/.opam/5.3~beta1/lib/dyn -I /home/opam/.opam/5.3~beta1/lib/fiber -I /home/opam/.opam/5.3~beta1/lib/jsonrpc -I /home/opam/.opam/5.3~beta1/lib/lsp -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/analysis -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/commands -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/dot_protocol -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/extend -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/kernel -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_merlin_specific -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_parsing -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_typing -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/ocaml_utils -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/query_commands -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/query_protocol -I /home/opam/.opam/5.3~beta1/lib/merlin-lib/utils -I /home/opam/.opam/5.3~beta1/lib/ocaml/unix -I /home/opam/.opam/5.3~beta1/lib/ocamlc-loc -I /home/opam/.opam/5.3~beta1/lib/ocamlformat-rpc-lib -I /home/opam/.opam/5.3~beta1/lib/ordering -I /home/opam/.opam/5.3~beta1/lib/pp -I /home/opam/.opam/5.3~beta1/lib/ppx_yojson_conv_lib -I /home/opam/.opam/5.3~beta1/lib/re -I /home/opam/.opam/5.3~beta1/lib/spawn -I /home/opam/.opam/5.3~beta1/lib/stdune -I /home/opam/.opam/5.3~beta1/lib/stdune/filesystem_stubs -I /home/opam/.opam/5.3~beta1/lib/xdg -I /home/opam/.opam/5.3~beta1/lib/yojson -I lsp-fiber/src/.lsp_fiber.objs/byte -I lsp-fiber/src/.lsp_fiber.objs/native -I ocaml-lsp-server/vendor/cmarkit/.cmarkit.objs/byte -I ocaml-lsp-server/vendor/cmarkit/.cmarkit.objs/native -I ocaml-lsp-server/vendor/lev-fiber-csexp/.lev_fiber_csexp.objs/byte -I ocaml-lsp-server/vendor/lev-fiber-csexp/.lev_fiber_csexp.objs/native -I ocaml-lsp-server/vendor/lev-fiber/.lev_fiber.objs/byte -I ocaml-lsp-server/vendor/lev-fiber/.lev_fiber.objs/native -I ocaml-lsp-server/vendor/lev/.lev.objs/byte -I ocaml-lsp-server/vendor/lev/.lev.objs/native -I vendor/odoc-parser/src/.odoc_parser.objs/byte -I vendor/odoc-parser/src/.odoc_parser.objs/native -intf-suffix .ml -no-alias-deps -open Ocaml_lsp_server__ -o ocaml-lsp-server/src/.ocaml_lsp_server.objs/native/ocaml_lsp_server__Semantic_highlighting.cmx -c -impl ocaml-lsp-server/src/semantic_highlighting.ml)
# File "ocaml-lsp-server/src/semantic_highlighting.ml", line 512, characters 8-34:
# 512 |       | Parsetree.Pconst_integer _ | Pconst_float _ -> Token_type.of_builtin Number
#               ^^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This pattern should not be a constructor, the expected type is
#        "Ocaml_lsp_server__.Import.Parsetree.constant"
```

**Comment by mseri** [comment]
A lower bound needs fixing:
```

#=== ERROR while compiling ocaml-lsp-server.1.20.0 ============================#
# context              2.4.0~alpha1~dev | linux/x86_64 | ocaml-base-compiler.5.2.1 | pinned(https://github.com/ocaml/ocaml-lsp/releases/download/1.20.0/lsp-1.20.0.tbz)
# path                 ~/.opam/5.2/.opam-switch/build/ocaml-lsp-server.1.20.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ocaml-lsp-server -j 39 @install
# exit-code            1
# env-file             ~/.opam/log/ocaml-lsp-server-7-d968f1.env
# output-file          ~/.opam/log/ocaml-lsp-server-7-d968f1.out
### output ###
# (cd _build/default && /home/opam/.opam/5.2/bin/ocamlc.opt -w -40 -alert -unstable -w -58 -g -bin-annot -I ocaml-lsp-server/src/.ocaml_lsp_server.objs/byte -I /home/opam/.opam/5.2/lib/base -I /home/opam/.opam/5.2/lib/chrome-trace -I /home/opam/.opam/5.2/lib/csexp -I /home/opam/.opam/5.2/lib/dune-build-info -I /home/opam/.opam/5.2/lib/dune-rpc -I /home/opam/.opam/5.2/lib/dune-rpc/private -I /home/opam/.opam/5.2/lib/dyn -I /home/opam/.opam/5.2/lib/fiber -I /home/opam/.opam/5.2/lib/jsonrpc -I /home/opam/.opam/5.2/lib/lsp -I /home/opam/.opam/5.2/lib/merlin-lib/analysis -I /home/opam/.opam/5.2/lib/merlin-lib/commands -I /home/opam/.opam/5.2/lib/merlin-lib/dot_protocol -I /home/opam/.opam/5.2/lib/merlin-lib/extend -I /home/opam/.opam/5.2/lib/merlin-lib/kernel -I /home/opam/.opam/5.2/lib/merlin-lib/ocaml_merlin_specific -I /home/opam/.opam/5.2/lib/merlin-lib/ocaml_parsing -I /home/opam/.opam/5.2/lib/merlin-lib/ocaml_typing -I /home/opam/.opam/5.2/lib/merlin-lib/ocaml_utils -I /home/opam/.opam/5.2/lib/merlin-lib/query_commands -I /home/opam/.opam/5.2/lib/merlin-lib/query_protocol -I /home/opam/.opam/5.2/lib/merlin-lib/utils -I /home/opam/.opam/5.2/lib/ocaml/unix -I /home/opam/.opam/5.2/lib/ocamlc-loc -I /home/opam/.opam/5.2/lib/ocamlformat-rpc-lib -I /home/opam/.opam/5.2/lib/ordering -I /home/opam/.opam/5.2/lib/pp -I /home/opam/.opam/5.2/lib/ppx_yojson_conv_lib -I /home/opam/.opam/5.2/lib/re -I /home/opam/.opam/5.2/lib/spawn -I /home/opam/.opam/5.2/lib/stdune -I /home/opam/.opam/5.2/lib/stdune/filesystem_stubs -I /home/opam/.opam/5.2/lib/xdg -I /home/opam/.opam/5.2/lib/yojson -I lsp-fiber/src/.lsp_fiber.objs/byte -I ocaml-lsp-server/vendor/cmarkit/.cmarkit.objs/byte -I ocaml-lsp-server/vendor/lev-fiber-csexp/.lev_fiber_csexp.objs/byte -I ocaml-lsp-server/vendor/lev-fiber/.lev_fiber.objs/byte -I ocaml-lsp-server/vendor/lev/.lev.objs/byte -I vendor/odoc-parser/src/.odoc_parser.objs/byte -no-alias-deps -open Ocaml_lsp_server__ -o ocaml-lsp-server/src/.ocaml_lsp_server.objs/byte/ocaml_lsp_server__Import.cmo -c -impl ocaml-lsp-server/src/import.ml)
# File "ocaml-lsp-server/src/import.ml", line 33, characters 17-18:
# 33 |         Int.incr i
#                       ^
# Error: This expression has type "int/2 Base.ref" = "int/2 Stdlib.ref"
#        but an expression was expected of type
#          "Base.Int.t Base__.Import.ref" = "Base.Int.t Caml.ref"
#        File "_none_", line 1:
#          Definition of type "int/2"
# (cd _build/default/ocaml-lsp-server/vendor/lev && /usr/bin/gcc -O2 -fno-strict-aliasing -fwrapv -pthread -fPIC -pthread -D_FILE_OFFSET_BITS=64 -g -I /home/opam/.opam/5.2/lib/ocaml -I /home/opam/.opam/5.2/lib/ocaml/unix -o ev.o -c ev.c)
# submodules/lev/lev/vendor/ev.c:2143:31: warning: 'ev_default_loop_ptr' initialized and declared 'extern'
```

**Comment by voodoos** [comment]
Thanks @mseri. Should be good with base >= 0.16.0.

(I also added a release compatible with 4.14, so we should wait for the CI)

**Comment by avsm** [comment]
All passed the CI! Thanks

**Pull request merged**